runtime/cudaq/utils/details/impls/xtensor_impl.cpp

```diff
@@ -0,0 +1,163 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2024 NVIDIA Corporation & Affiliates.                         *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "cudaq/utils/details/tensor_impl.h"
+#include "cudaq/utils/type_traits.h"
+
+#include <xtensor/xadapt.hpp>
+#include <xtensor/xio.hpp>
+
+#include <fmt/ranges.h>
+
+namespace cudaqx {
+
+/// @brief An implementation of tensor_impl using xtensor library
+template <typename Scalar>
+class xtensor : public cudaqx::details::tensor_impl<Scalar> {
+private:
+  Scalar *m_data = nullptr;         ///< Pointer to the tensor data
+  std::vector<std::size_t> m_shape; ///< Shape of the tensor
+  bool ownsData = true; ///< Flag indicating if this object owns the data
+
+  /// @brief Check if the given indices are valid for this tensor
+  /// @param idxs Vector of indices to check
+  /// @return true if indices are valid, false otherwise
+  bool validIndices(const std::vector<std::size_t> &idxs) const {
+    if (idxs.size() != m_shape.size())
+      return false;
+    for (std::size_t dim = 0; auto idx : idxs)
+      if (idx < 0 || idx >= m_shape[dim++])
+        return false;
+    return true;
+  }
+
+public:
+  /// @brief Constructor for xtensor
+  /// @param d Pointer to the tensor data
+  /// @param s Shape of the tensor
+  xtensor(const Scalar *d, const std::vector<std::size_t> &s)
+      : m_data(const_cast<Scalar *>(d)), m_shape(s) {}
```

Comment:
This takes ownership of the data, and if that is intentional, the docs should be updated to reflect that. (Otherwise, users may be tempted to pass `theirVec.data()` into the constructor and then be confused when a double free (or something worse) happens.